include/circt/Support/BuilderUtils.h

```diff
@@ -0,0 +1,45 @@
+//===- BuilderUtils.h - Operation builder utilities -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_BUILDERUTILS_H
+#define CIRCT_SUPPORT_BUILDERUTILS_H
+
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace circt {
+
+/// A helper union that can represent a `StringAttr`, `StringRef`, or `Twine`.
+/// It is intended to be used as arguments to an op's `build` function. This
+/// allows a single builder to accept any flavor value for a string attribute.
+/// The `get` function can then be used to obtain a `StringAttr` from any of the
+/// possible variants `StringAttrOrRef` can take.
+class StringAttrOrRef {
```

Comment:
This is a Pandora's box. Can't we get away with two constructors for wire instead? One to take two StringAttrs and one to take two Twines and re-write the instantiation sites to eliminate mixed uses?

Reply:
The problem is that we use this `StringRef` and `StringAttr` duality all over the place, and there seems to be a definite desire to have builders that accept `StringRef`. But always having the full canonical list of combinations seems like an anti-pattern. Ideally we'd be able to upstream some of these helpers to MLIR eventually, also for arrays and friends, such that TableGen can directly generate them.